src/inference/dreamEubo.ad.js

```diff
@@ -0,0 +1,189 @@
+    // Records can be created by using dreamSample.
+    this.records = this.opts.records;
+    this.currRecord;
+
+    this.params = params;
+    this.cont = cont;
+
+    this.wpplFn = wpplFn;
+    this.s = s;
+    this.a = a;
+
+    // The global model includes anything outside of any mapData (level 0) 
+    // The local model includes anything inside of some mapData (level 1+)
+    this.mapDataNestingLevel = 0;
+
+    this.coroutine = env.coroutine;
+    env.coroutine = this;
+  }
+
+  function isInsideMapData() {
+    return this.mapDataNestingLevel > 0;
+  }
+
+  function checkScoreIsFinite(score, source) {
+    var _score = ad.value(score);
+    if (!isFinite(_score)) { // Also catches NaN.
+
+      var msg = 'DREAM: The score of the previous sample under the ' + source;
+      msg += ' program was ' + _score + '.';
+      if (_.isNaN(_score)) {
+        msg += ' Reducing the step size may help.';
+      }
+      throw new Error(msg);
+    }
+  }
+
+  DREAM.prototype = {
+
+    // Computes the estimation of the objective EUBO by averaging the
+    // gradients for all the records provided to the module through params,
+    // so as to get an estimated expectation.
+    run: function() {
+
+      var dream = 0;
+      var grad = {};
+
+      return util.cpsForEach(
+
+          // Body.
+          function(record, i, records, next) {
+            return this.estimateGradient(record, function(g, dream_i) {
+              paramStruct.addEq(grad, g); // Accumulates gradient estimates.
+              dream += dream_i;
+              return next();
+            });
+          }.bind(this),
+
+          // Continuation.
+          function() {
+            paramStruct.divEq(grad, this.records.length);
+            dream /= this.records.length; // Averages estimations.
+            env.coroutine = this.coroutine;
+            return this.cont(grad, dream);
+          }.bind(this),
+
+          this.records);
+
+    },
+
+    // Computes a single sample estimate of the gradient,
+    // which in case of EUBO is the minus guide scoring on
+    // the target recorded samples.
+    estimateGradient: function(record, cont) {
+
+      // Makes record available throughout the module.
+      this.currRecord = record;
+
+      // paramsSeen tracks the AD nodes of all parameters seen during
+      // a single execution. These are the parameters for which
+      // gradients will be computed.
+      this.paramsSeen = {};
+      this.logq = 0;
+
+      return this.wpplFn(_.clone(this.s), function() {
+
+        var objective = -this.logq;
+
+        if (ad.isLifted(objective)) { // Handles programs with zero random choices.
+          objective.backprop();
+        }
+
+        var grads = _.mapObject(this.paramsSeen, function(params) {
+          return params.map(ad.derivative);
+        });
+
+        var logp = ad.value(this.currRecord.trace.score);
+        var logq = ad.value(this.logq);
+        return cont(grads, logp - logq);
```

Comment:
Since we only accumulate the score of local choices in `this.logq`, I wonder if `logp` should also be the sum of the scores from the local choices? (i.e. not include the global choices, as I assume the trace score does.)